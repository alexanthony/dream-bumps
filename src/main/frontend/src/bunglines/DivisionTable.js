// @flow
import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import {
  Table,
  TableBody,
  TableHeader,
  TableHeaderColumn,
  TableRow,
  TableRowColumn
} from 'material-ui/Table';
import FlatButton from 'material-ui/FlatButton';
import RaisedButton from 'material-ui/RaisedButton';
import { Link } from 'react-router';
import UpArrow from 'material-ui/svg-icons/navigation/arrow-upward';
import RightArrow from 'material-ui/svg-icons/navigation/arrow-forward';
import MediaQuery from 'react-responsive';

import {
  getCurrentUserId,
  getUserCrewRanking,
  getBuySex,
  getBuyMemberId,
  getNumberOfCrews,
  getCrewsPerDivision
} from '../rootReducer';
import { doBuyRower } from '../usercrews/actions';
import {
  calculateRowOverDividendForPosition,
  calculateBumpDividend
} from '../utils/dividends';
import { formatMoney } from '../utils/maths';
/**
 * DivisionTable
 */
const DivisionTable = props => (
  <Table selectable={false}>
    <TableHeader adjustForCheckbox={false} displaySelectAll={false}>
      <TableRow>
        <TableHeaderColumn>
          <h2>{`Division ${parseInt(props.division, 10) + 1}`}</h2>
        </TableHeaderColumn>
      </TableRow>
    </TableHeader>
    <TableBody displayRowCheckbox={false}>
      {props.crews.map(crew => (
        <TableRow key={crew.id}>
          <MediaQuery minWidth={768}>
            {matches => (
              <TableRowColumn
                style={{
                  paddingLeft: 0,
                  paddingRight: 0,
                  width: matches ? '45px' : '26px'
                }}
              >
                {crew.position}
              </TableRowColumn>
            )}
          </MediaQuery>
          <MediaQuery minWidth={768}>
            {matches => (
              <TableRowColumn
                style={{
                  paddingLeft: matches ? '24px' : '0px'
                }}
              >
                <FlatButton
                  containerElement={<Link to={`/crews/${crew.id}`} />}
                  label={crew.name}
                  labelStyle={{ textTransform: '', fontSize: '13px' }}
                  style={{
                    paddingLeft: 0,
                    textAlign: '',
                    marginLeft: '-16px',
                    minWidth: '80px'
                  }}
                />
              </TableRowColumn>
            )}
          </MediaQuery>
          {props.buySex === props.sex
            ? <MediaQuery minWidth={768}>
                {matches => (
                  <TableRowColumn
                    style={
                      !matches && {
                        width: '68px',
                        paddingLeft: '0px',
                        paddingRight: '8px'
                      }
                    }
                  >
                    <RaisedButton
                      style={
                        !matches && {
                          minWidth: '68px',
                          paddingLeft: '0px',
                          paddingRight: '0px'
                        }
                      }
                      label="Buy"
                      primary
                      disabled={crew.price > props.crewRanking.cash}
                      onClick={e =>
                        props.doBuyRower(crew.id, props.buyMemberId)}
                    />
                  </TableRowColumn>
                )}
              </MediaQuery>
            : null}
          <MediaQuery minWidth={768}>
            {matches => (
              <TableRowColumn
                style={{
                  width: matches ? '60px' : '40px',
                  paddingRight: matches ? '8px' : '4px',
                  paddingLeft: matches ? '8px' : '4px'
                }}
              >
                {formatMoney(crew.price)}
              </TableRowColumn>
            )}
          </MediaQuery>
          <MediaQuery minWidth={768}>
            {matches => (
              <TableRowColumn
                style={{
                  width: matches ? '80px' : '70px',
                  paddingRight: matches ? '8px' : '4px',
                  paddingLeft: matches ? '8px' : '4px',
                  display: matches || props.buySex !== props.sex
                    ? 'table-cell'
                    : 'none'
                }}
              >
                {crew.position > 1 && <UpArrow />}
                {crew.position > 1 &&
                  calculateBumpDividend(
                    crew.position,
                    props.numberOfCrews,
                    props.crewsPerDivision,
                    props.day
                  )}
              </TableRowColumn>
            )}
          </MediaQuery>
          <MediaQuery minWidth={768}>
            {matches => (
              <TableRowColumn
                style={{
                  width: matches ? '80px' : '70px',
                  paddingRight: matches ? '8px' : '4px',
                  paddingLeft: matches ? '8px' : '4px',
                  display: matches || props.buySex !== props.sex
                    ? 'table-cell'
                    : 'none'
                }}
              >
                <RightArrow />
                {' '}
                {calculateRowOverDividendForPosition(crew.position, props.day)}
              </TableRowColumn>
            )}
          </MediaQuery>
        </TableRow>
      ))}
    </TableBody>
  </Table>
);

DivisionTable.propTypes = {
  sex: PropTypes.string,
  division: PropTypes.string,
  crews: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      position: PropTypes.number,
      name: PropTypes.string
    })
  ),
  buySex: PropTypes.string,
  crewRanking: PropTypes.shape({
    cash: PropTypes.number
  }),
  doBuyRower: PropTypes.func,
  buyMemberId: PropTypes.number,
  crewsPerDivision: PropTypes.number,
  numberOfCrews: PropTypes.number,
  day: PropTypes.number
};

export const mapStateToProps = (state: Object, ownProps: Object) => {
  const mapping = {
    // crews: getCrewsForDivision(state)(ownProps.sex, ownProps.division),
    currentUserId: getCurrentUserId(state),
    buySex: getBuySex(state),
    buyMemberId: getBuyMemberId(state),
    // Give flow a hand - objects should be sealed when declared
    crewRanking: {},
    crewsPerDivision: 0,
    numberOfCrews: 0
  };
  mapping.crewRanking = getUserCrewRanking(state)(
    mapping.currentUserId,
    ownProps.sex
  ) || {};
  mapping.crewsPerDivision = getCrewsPerDivision(state)();
  mapping.numberOfCrews = getNumberOfCrews(state)(ownProps.sex);
  return mapping;
};

export { DivisionTable };

export default connect(mapStateToProps, { doBuyRower })(DivisionTable);
