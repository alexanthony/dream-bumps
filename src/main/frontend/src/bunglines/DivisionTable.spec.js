import React from 'react';
import { shallow } from 'enzyme';
import { DivisionTable, mapStateToProps } from './DivisionTable';
import toJson from 'enzyme-to-json';

test('Renders properly not buy', () => {
  const crews = [
    {id: 1, name: 'Crew 1', position: 1, price: 123.4567},
    {id: 3, name: 'Crew 2', position: 2, price: 153.4}
   ]
  const component = shallow(
    <DivisionTable
      sex="male"
      division="2"
      crews={crews}
      buySex="female"
      day={2}
      crewRanking={{cash: 130}}
      numberOfCrews={92}
      crewsPerDivision={13}
    />
  );
  expect(component).toBeDefined();
  expect(toJson(component)).toMatchSnapshot()
});

test('Renders properly buy', () => {
  const crews = [
    {id: 1, name: 'Crew 1', position: 1, price: 123.4567},
    {id: 3, name: 'Crew 2', position: 2, price: 153.4}
   ]
  const component = shallow(
    <DivisionTable
      sex="male"
      division="2"
      crews={crews}
      buySex="male"
      day={2}
      crewRanking={{cash: 130}}
      numberOfCrews={92}
      crewsPerDivision={13}
    />
  );
  expect(component).toBeDefined();
  expect(toJson(component)).toMatchSnapshot()
});

test('mapStateToProps', () => {
  const reduxState = {
    auth: {
      authenticated: true,
      currentUser: {
        authorities: ['ROLE_USER'],
        id: 2
      }
    },
    config: {
      config: {
        crewsPerDivision: 3,
        mensCrews: 13,
        womensCrews: 10
      }
    },
    usercrews: {
      purchase: {
        buyMember: 5,
        buySex: 'male'},
      rankings: {
        byId: {
          6: {id: 6, cash: 123},
          7: {id: 7, cash: 456},
          8: {id: 8, cash: 789},
          9: {id: 9, cash: 987}
        },
        byUserId: {
          1: {
            male: 6,
            female: 7
          },
          2: {
            male: 8,
            female: 9
          }
        }
      }
    }
  };

  const ownProps = {sex: 'male'};

  const componentStateProps = mapStateToProps(reduxState, ownProps);

  expect(componentStateProps).toEqual({
    currentUserId: 2,
    buySex: 'male',
    buyMemberId: 5,
    crewRanking: {id: 8, cash: 789},
    crewsPerDivision: 3,
    numberOfCrews: 13
  });
});
