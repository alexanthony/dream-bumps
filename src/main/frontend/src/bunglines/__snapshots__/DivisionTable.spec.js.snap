exports[`test Renders properly buy 1`] = `
<Table
  allRowsSelected={false}
  fixedFooter={true}
  fixedHeader={true}
  height="inherit"
  multiSelectable={false}
  selectable={false}>
  <TableHeader
    adjustForCheckbox={false}
    displaySelectAll={false}
    enableSelectAll={true}
    selectAllSelected={false}>
    <TableRow
      displayBorder={true}
      hoverable={false}
      hovered={false}
      selectable={true}
      selected={false}
      striped={false}>
      <TableHeaderColumn>
        <h2>
          Division 3
        </h2>
      </TableHeaderColumn>
    </TableRow>
  </TableHeader>
  <TableBody
    allRowsSelected={false}
    deselectOnClickaway={true}
    displayRowCheckbox={false}
    multiSelectable={false}
    preScanRows={true}
    selectable={true}
    style={Object {}}>
    <TableRow
      displayBorder={true}
      hoverable={false}
      hovered={false}
      selectable={true}
      selected={false}
      striped={false}>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "paddingLeft": 0,
            "paddingRight": 0,
            "width": "45px",
          }
        }>
        1
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}>
        <FlatButton
          containerElement={
            <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to="/crews/1" />
          }
          disabled={false}
          label="Crew 1"
          labelPosition="after"
          labelStyle={
            Object {
              "fontSize": "13px",
              "textTransform": "",
            }
          }
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          primary={false}
          secondary={false}
          style={
            Object {
              "marginLeft": "-16px",
              "minWidth": "80px",
              "paddingLeft": 0,
              "textAlign": "",
            }
          } />
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}>
        <RaisedButton
          disabled={false}
          fullWidth={false}
          label="Buy"
          labelPosition="after"
          onClick={[Function]}
          primary={true}
          secondary={false} />
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        }>
        123.46
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        } />
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        }>
        <NavigationArrowForward />
         
        90.00
      </TableRowColumn>
    </TableRow>
    <TableRow
      displayBorder={true}
      hoverable={false}
      hovered={false}
      selectable={true}
      selected={false}
      striped={false}>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "paddingLeft": 0,
            "paddingRight": 0,
            "width": "45px",
          }
        }>
        2
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}>
        <FlatButton
          containerElement={
            <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to="/crews/3" />
          }
          disabled={false}
          label="Crew 2"
          labelPosition="after"
          labelStyle={
            Object {
              "fontSize": "13px",
              "textTransform": "",
            }
          }
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          primary={false}
          secondary={false}
          style={
            Object {
              "marginLeft": "-16px",
              "minWidth": "80px",
              "paddingLeft": 0,
              "textAlign": "",
            }
          } />
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}>
        <RaisedButton
          disabled={true}
          fullWidth={false}
          label="Buy"
          labelPosition="after"
          onClick={[Function]}
          primary={true}
          secondary={false} />
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        }>
        153.40
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        }>
        <NavigationArrowUpward />
        228.83
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        }>
        <NavigationArrowForward />
         
        76.28
      </TableRowColumn>
    </TableRow>
  </TableBody>
</Table>
`;

exports[`test Renders properly not buy 1`] = `
<Table
  allRowsSelected={false}
  fixedFooter={true}
  fixedHeader={true}
  height="inherit"
  multiSelectable={false}
  selectable={false}>
  <TableHeader
    adjustForCheckbox={false}
    displaySelectAll={false}
    enableSelectAll={true}
    selectAllSelected={false}>
    <TableRow
      displayBorder={true}
      hoverable={false}
      hovered={false}
      selectable={true}
      selected={false}
      striped={false}>
      <TableHeaderColumn>
        <h2>
          Division 3
        </h2>
      </TableHeaderColumn>
    </TableRow>
  </TableHeader>
  <TableBody
    allRowsSelected={false}
    deselectOnClickaway={true}
    displayRowCheckbox={false}
    multiSelectable={false}
    preScanRows={true}
    selectable={true}
    style={Object {}}>
    <TableRow
      displayBorder={true}
      hoverable={false}
      hovered={false}
      selectable={true}
      selected={false}
      striped={false}>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "paddingLeft": 0,
            "paddingRight": 0,
            "width": "45px",
          }
        }>
        1
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}>
        <FlatButton
          containerElement={
            <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to="/crews/1" />
          }
          disabled={false}
          label="Crew 1"
          labelPosition="after"
          labelStyle={
            Object {
              "fontSize": "13px",
              "textTransform": "",
            }
          }
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          primary={false}
          secondary={false}
          style={
            Object {
              "marginLeft": "-16px",
              "minWidth": "80px",
              "paddingLeft": 0,
              "textAlign": "",
            }
          } />
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        }>
        123.46
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        } />
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        }>
        <NavigationArrowForward />
         
        90.00
      </TableRowColumn>
    </TableRow>
    <TableRow
      displayBorder={true}
      hoverable={false}
      hovered={false}
      selectable={true}
      selected={false}
      striped={false}>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "paddingLeft": 0,
            "paddingRight": 0,
            "width": "45px",
          }
        }>
        2
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}>
        <FlatButton
          containerElement={
            <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to="/crews/3" />
          }
          disabled={false}
          label="Crew 2"
          labelPosition="after"
          labelStyle={
            Object {
              "fontSize": "13px",
              "textTransform": "",
            }
          }
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          primary={false}
          secondary={false}
          style={
            Object {
              "marginLeft": "-16px",
              "minWidth": "80px",
              "paddingLeft": 0,
              "textAlign": "",
            }
          } />
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        }>
        153.40
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        }>
        <NavigationArrowUpward />
        228.83
      </TableRowColumn>
      <TableRowColumn
        hoverable={false}
        style={
          Object {
            "width": "100px",
          }
        }>
        <NavigationArrowForward />
         
        76.28
      </TableRowColumn>
    </TableRow>
  </TableBody>
</Table>
`;
